File: backend/atlas/analysis/latent.py
"""Patrones ocultos: codificador de grafo con una capa de atención (tipo transformer sobre el vecindario) entrenado
para predecir enlaces. Espejo en numpy de frontend/src/analysis/latent.ts (misma arquitectura y objetivo; los
generadores aleatorios difieren, por lo que los pares concretos pueden variar entre navegador y CLI).

Salidas: relaciones latentes (pares sin enlace con alta probabilidad), comunidades (k-medias por coseno),
flujos de fechas y posibles reubicaciones. Solo metadata (regla M4); todo es hipótesis `inferred`/`proposed`.
"""
from __future__ import annotations

import time

import numpy as np

from ..embeddings.hashing import HashProvider
from ..text import tokens

STOP = {"de", "del", "la", "el", "los", "las", "en", "al", "y", "o", "con", "por", "a", "un", "una", "the", "of", "para"}


def node_text(n: dict) -> str:
    f = n.get("facets") or {}
    return ". ".join(str(x) for x in [n.get("name", ""), n.get("description") or n.get("note") or "", f.get("data_type") or "", f.get("info_domain") or "", f.get("element_kind") or ""] if x)


def train_latent(agf: dict, dim: int = 32, epochs: int = 120, lr: float = 0.03, seed: int = 7, negatives: int = 2,
                 top_pairs: int = 40, min_score: float = 0.8, communities: int | None = None) -> dict:
    t0 = time.time()
    rng = np.random.default_rng(seed)
    nodes = [n for n in agf["nodes"] if n["kind"] in ("element", "section")]
    idx = {n["id"]: i for i, n in enumerate(nodes)}
    N = len(nodes)
    hp = HashProvider()
    X = np.array([hp.embed_one(node_text(n), "passage") for n in nodes], dtype=float)   # N × F
    F = X.shape[1]
    edges, holdout = [], []
    for e in agf["edges"]:
        if e.get("status") == "rejected":
            continue
        a, b = idx.get(e["source"]), idx.get(e["target"])
        if a is None or b is None or a == b:
            continue
        w = 0.5 if e["kind"] == "relates" else 1.0
        if e["kind"] in ("same_as", "precedes") and rng.random() < 0.15:
            holdout.append((a, b)); continue
        edges.append((a, b, w))
    adj = [[] for _ in range(N)]
    adjset = set()
    for a, b, _ in edges:
        adj[a].append(b); adj[b].append(a); adjset.add((a, b)); adjset.add((b, a))
    known = set(adjset)
    for a, b in holdout:
        known.add((a, b)); known.add((b, a))
    nb = [[i] + adj[i] for i in range(N)]
    D = dim
    Win = rng.uniform(-0.3, 0.3, (D, F)); Q = rng.uniform(-0.3, 0.3, (D, D)); K = rng.uniform(-0.3, 0.3, (D, D)); V = rng.uniform(-0.3, 0.3, (D, D))
    scale = 1 / np.sqrt(D)
    sigmoid = lambda x: 1 / (1 + np.exp(-x))  # noqa: E731

    def forward():
        H = X @ Win.T; QH = H @ Q.T; KH = H @ K.T; VH = H @ V.T
        Z = np.zeros_like(H); AL = []
        for i in range(N):
            js = nb[i]
            e = (KH[js] @ QH[i]) * scale
            e = np.exp(e - e.max()); al = e / e.sum()
            Z[i] = H[i] + al @ VH[js]
            AL.append(al)
        return H, QH, KH, VH, Z, AL

    loss_start = loss_end = 0.0
    E = np.array([(a, b) for a, b, _ in edges], dtype=int); W = np.array([w for _, _, w in edges])
    for ep in range(epochs):
        H, QH, KH, VH, Z, AL = forward()
        dZ = np.zeros_like(Z); loss = 0.0; count = 0
        # positivos
        s = np.einsum("ij,ij->i", Z[E[:, 0]], Z[E[:, 1]]); p = sigmoid(s)
        loss += float(-(W * np.log(p + 1e-9)).sum()); count += len(E)
        g = W * (p - 1)
        np.add.at(dZ, E[:, 0], g[:, None] * Z[E[:, 1]]); np.add.at(dZ, E[:, 1], g[:, None] * Z[E[:, 0]])
        # negativos
        for _ in range(negatives):
            c = rng.integers(0, N, len(E))
            ok = np.array([ci != a and (a, ci) not in adjset for a, ci in zip(E[:, 0], c)])
            if not ok.any():
                continue
            a_, c_, w_ = E[ok, 0], c[ok], W[ok] / negatives
            s = np.einsum("ij,ij->i", Z[a_], Z[c_]); p = sigmoid(s)
            loss += float(-(w_ * np.log(1 - p + 1e-9)).sum()); count += len(a_)
            g = w_ * p
            np.add.at(dZ, a_, g[:, None] * Z[c_]); np.add.at(dZ, c_, g[:, None] * Z[a_])
        loss /= max(1, count)
        if ep == 0:
            loss_start = loss
        loss_end = loss
        dH = dZ.copy(); dQH = np.zeros_like(H); dKH = np.zeros_like(H); dVH = np.zeros_like(H)
        for i in range(N):
            js = nb[i]; al = AL[i]; gz = dZ[i]
            dal = VH[js] @ gz
            dVH[js] += al[:, None] * gz
            de = al * (dal - float(al @ dal)) * scale
            dQH[i] += de @ KH[js]
            dKH[js] += de[:, None] * QH[i]
        dQ = dQH.T @ H; dK = dKH.T @ H; dV = dVH.T @ H
        dH += dQH @ Q + dKH @ K + dVH @ V
        dWin = dH.T @ X
        step = lr / max(1, count / 40)
        Win -= step * dWin; Q -= step * dQ; K -= step * dK; V -= step * dV
    H, QH, KH, VH, Z, AL = forward()
    score = lambda a, b: float(sigmoid(Z[a] @ Z[b]))  # noqa: E731
    auc = None
    if len(holdout) >= 3:
        wins = total = 0
        for a, b in holdout:
            for _ in range(20):
                c = int(rng.integers(0, N))
                if c == a or (a, c) in adjset:
                    continue
                total += 1; wins += score(a, b) > score(a, c)
        auc = wins / total if total else None
    els = [i for i, n in enumerate(nodes) if n["kind"] == "element"]
    cand = []
    for x in range(len(els)):
        for y in range(x + 1, len(els)):
            a, b = els[x], els[y]
            if (a, b) in known or nodes[a].get("parent_id") == nodes[b].get("parent_id"):
                continue
            s = score(a, b)
            if s < min_score:
                continue
            shared = [nodes[j]["id"] for j in adj[a] if j in adj[b] and nodes[j]["kind"] == "element"]
            ta = set(tokens(nodes[a]["name"])) - STOP; tb = [t for t in tokens(nodes[b]["name"]) if t in ta and t not in STOP]
            why = "; ".join(w for w in [f"{len(shared)} vecino(s) en común" if shared else "", f"comparten «{'», «'.join(tb[:3])}»" if tb else "",
                                        "mismo tipo de dato" if (nodes[a].get("facets") or {}).get("data_type") == (nodes[b].get("facets") or {}).get("data_type") else "mismo contexto de sección"] if w)
            cand.append({"a": nodes[a]["id"], "b": nodes[b]["id"], "score": round(s, 3), "shared": shared, "why": why or "cercanía en el espacio aprendido",
                         "same_source": nodes[a].get("source_id") == nodes[b].get("source_id")})
    cand.sort(key=lambda c: -c["score"])
    pairs = cand[:top_pairs]
    # comunidades
    U = Z[els] / (np.linalg.norm(Z[els], axis=1, keepdims=True) + 1e-9)
    k = communities or max(2, min(8, round(np.sqrt(len(els) / 2))))
    centers = U[:k].copy(); assign = np.zeros(len(els), dtype=int)
    for _ in range(15):
        assign = (U @ centers.T).argmax(axis=1)
        for j in range(k):
            m = U[assign == j]
            if len(m):
                c = m.sum(axis=0); centers[j] = c / (np.linalg.norm(c) + 1e-9)
    comms = []
    for j in range(k):
        members = [els[i] for i in range(len(els)) if assign[i] == j]
        if len(members) < 2:
            continue
        freq: dict[str, int] = {}
        for i in members:
            for t in set(t for t in tokens(nodes[i]["name"]) if t not in STOP and len(t) > 2):
                freq[t] = freq.get(t, 0) + 1
        top = [t for t, _ in sorted(freq.items(), key=lambda x: -x[1])[:3]]
        coh = float(np.mean([U[els.index(i)] @ centers[j] for i in members]))
        comms.append({"id": j, "label": " · ".join(top) or f"grupo {j + 1}", "members": [nodes[i]["id"] for i in members], "cohesion": round(coh, 3)})
    comms.sort(key=lambda c: -len(c["members"]))
    return {"pairs": pairs, "communities": comms,
            "training": {"nodes": N, "edges": len(edges), "epochs": epochs, "lossStart": round(loss_start, 4), "lossEnd": round(loss_end, 4),
                         "holdoutAuc": None if auc is None else round(auc, 3), "seed": seed, "dim": dim, "ms": int((time.time() - t0) * 1000)}}


def as_edges(pairs: list[dict], start: int = 0) -> list[dict]:
    """Convierte pares latentes en aristas `relates` (predicate latent) propuestas."""
    return [{"id": f"lt{start + i + 1}", "kind": "relates", "source": p["a"], "target": p["b"], "predicate": "latent", "layer": "latent",
             "origin": "inferred", "status": "proposed", "confidence": p["score"], "rationale": "modelo de atención: " + p["why"]} for i, p in enumerate(pairs)]

File: frontend/src/analysis/latent.ts
/** Patrones ocultos: un codificador de grafo con atención (una capa tipo transformer sobre el vecindario de
 *  cada nodo) entrenado en el navegador para predecir enlaces. Con lo aprendido se proponen:
 *    - relaciones latentes: pares no conectados con alta probabilidad de enlace;
 *    - comunidades: grupos de datos que el modelo ve juntos, con nombre por sus palabras frecuentes;
 *    - flujos: cadenas de fechas ordenadas por el documento y reforzadas por enlaces predichos;
 *    - reubicaciones: datos que el modelo sitúa más cerca de otra sección que de la suya.
 *  Límites: trabaja solo con metadata (nombres, facetas y estructura; nunca filas, regla M4). Todo resultado es
 *  hipótesis (`origin: inferred`, `status: proposed`) con puntaje y justificación; requiere validación humana.
 *  Determinista para una semilla dada. Espejo en backend/atlas/analysis/latent.py. */
import { norm, tokens } from "../lib/text";
import { hashOne } from "../search/hash";
import type { Agf, AgfNode } from "../types";

export interface LatentOptions { dim?: number; epochs?: number; lr?: number; seed?: number; negatives?: number; topPairs?: number; minScore?: number; communities?: number }
export interface LatentPair { a: string; b: string; score: number; shared: string[]; why: string; sameSource: boolean }
export interface Community { id: number; label: string; members: string[]; cohesion: number }
export interface Flow { steps: { id: string; predicted: boolean; score: number }[]; label: string }
export interface Relocation { id: string; from: string; to: string; margin: number }
export interface LatentResult {
  pairs: LatentPair[]; communities: Community[]; flows: Flow[]; relocations: Relocation[];
  training: { nodes: number; edges: number; epochs: number; lossStart: number; lossEnd: number; holdoutAuc: number | null; seed: number; dim: number; ms: number };
}

const STOP = new Set(["de", "del", "la", "el", "los", "las", "en", "al", "y", "o", "con", "por", "a", "un", "una", "the", "of", "para"]);

/** Generador determinista (mulberry32). */
function rng(seed: number) { let t = seed >>> 0; return () => { t += 0x6d2b79f5; let r = Math.imul(t ^ (t >>> 15), 1 | t); r ^= r + Math.imul(r ^ (r >>> 7), 61 | r); return ((r ^ (r >>> 14)) >>> 0) / 4294967296; }; }
const sigmoid = (x: number) => 1 / (1 + Math.exp(-x));

export function nodeText(n: AgfNode): string {
  const f = n.facets || {};
  return [n.name, n.description || n.note || "", f.data_type || "", f.info_domain || "", f.element_kind || ""].filter(Boolean).join(". ");
}

export function trainLatent(agf: Agf, opt: LatentOptions = {}): LatentResult {
  const t0 = performance.now();
  const D = opt.dim ?? 32, EPOCHS = opt.epochs ?? 120, LR = opt.lr ?? 0.03, SEED = opt.seed ?? 7, NEG = opt.negatives ?? 2;
  const rand = rng(SEED);
  const nodes = agf.nodes.filter((n) => n.kind === "element" || n.kind === "section");
  const idx = new Map(nodes.map((n, i) => [n.id, i]));
  const N = nodes.length;
  const X = nodes.map((n) => hashOne(nodeText(n)));   // 256-d léxico, norma 1
  const F = X[0]?.length || 256;
  // aristas (no dirigidas) con peso; se separa un 15 % de same_as/precedes como prueba ciega
  const edges: [number, number, number][] = [];
  const holdout: [number, number][] = [];
  for (const e of agf.edges) {
    if (e.status === "rejected") continue;
    const a = idx.get(e.source), b = idx.get(e.target);
    if (a === undefined || b === undefined || a === b) continue;
    const w = e.kind === "contains" ? 1 : e.kind === "relates" ? 0.5 : 1;
    if ((e.kind === "same_as" || e.kind === "precedes") && rand() < 0.15) { holdout.push([a, b]); continue; }
    edges.push([a, b, w]);
  }
  const adj: number[][] = Array.from({ length: N }, () => []);
  const adjSet = new Set<string>();
  for (const [a, b] of edges) { adj[a].push(b); adj[b].push(a); adjSet.add(a + "|" + b); adjSet.add(b + "|" + a); }
  const known = new Set(adjSet);   // toda relación registrada, incluidas las retenidas para la prueba ciega
  for (const [a, b] of holdout) { known.add(a + "|" + b); known.add(b + "|" + a); }
  // parámetros
  const init = (rows: number, cols: number, scale: number) => { const m = new Float64Array(rows * cols); for (let i = 0; i < m.length; i++) m[i] = (rand() * 2 - 1) * scale; return m; };
  const Win = init(D, F, 0.3), Q = init(D, D, 0.3), K = init(D, D, 0.3), V = init(D, D, 0.3);
  const matvec = (M: Float64Array, v: ArrayLike<number>, rows: number, cols: number) => { const o = new Float64Array(rows); for (let r = 0; r < rows; r++) { let s = 0; const off = r * cols; for (let c = 0; c < cols; c++) s += M[off + c] * v[c]; o[r] = s; } return o; };
  const matTvec = (M: Float64Array, v: Float64Array, rows: number, cols: number) => { const o = new Float64Array(cols); for (let r = 0; r < rows; r++) { const off = r * cols, vr = v[r]; if (!vr) continue; for (let c = 0; c < cols; c++) o[c] += M[off + c] * vr; } return o; };
  const outer = (G: Float64Array, a: Float64Array, b: ArrayLike<number>, rows: number, cols: number) => { for (let r = 0; r < rows; r++) { const ar = a[r]; if (!ar) continue; const off = r * cols; for (let c = 0; c < cols; c++) G[off + c] += ar * b[c]; } };
  const dot = (a: Float64Array, b: Float64Array) => { let s = 0; for (let i = 0; i < a.length; i++) s += a[i] * b[i]; return s; };
  const scale = 1 / Math.sqrt(D);

  let H: Float64Array[] = [], Z: Float64Array[] = [], QH: Float64Array[] = [], KH: Float64Array[] = [], VH: Float64Array[] = [], ALPHA: Float64Array[] = [];
  const forward = () => {
    H = X.map((x) => matvec(Win, x, D, F));
    QH = H.map((h) => matvec(Q, h, D, D)); KH = H.map((h) => matvec(K, h, D, D)); VH = H.map((h) => matvec(V, h, D, D));
    Z = []; ALPHA = [];
    for (let i = 0; i < N; i++) {
      const nb = [i, ...adj[i]];
      const e = nb.map((j) => dot(QH[i], KH[j]) * scale), m = Math.max(...e);
      const ex = e.map((v) => Math.exp(v - m)), s = ex.reduce((a, b) => a + b, 0);
      const alpha = new Float64Array(nb.length); for (let k = 0; k < nb.length; k++) alpha[k] = ex[k] / s;
      const z = new Float64Array(D);
      for (let k = 0; k < nb.length; k++) { const vh = VH[nb[k]], a = alpha[k]; for (let d = 0; d < D; d++) z[d] += a * vh[d]; }
      for (let d = 0; d < D; d++) z[d] += H[i][d];
      Z.push(z); ALPHA.push(alpha);
    }
  };
  let lossStart = 0, lossEnd = 0;
  for (let ep = 0; ep < EPOCHS; ep++) {
    forward();
    const dZ = Z.map(() => new Float64Array(D));
    let loss = 0, count = 0;
    const pair = (a: number, b: number, y: number, w: number) => {
      const s = dot(Z[a], Z[b]), p = sigmoid(s);
      loss += -w * (y ? Math.log(p + 1e-9) : Math.log(1 - p + 1e-9)); count++;
      const g = w * (p - y);
      for (let d = 0; d < D; d++) { dZ[a][d] += g * Z[b][d]; dZ[b][d] += g * Z[a][d]; }
    };
    for (const [a, b, w] of edges) {
      pair(a, b, 1, w);
      for (let k = 0; k < NEG; k++) { const c = Math.floor(rand() * N); if (c !== a && !adjSet.has(a + "|" + c)) pair(a, c, 0, w / NEG); }
    }
    loss /= Math.max(1, count);
    if (ep === 0) lossStart = loss; lossEnd = loss;
    // retropropagación
    const dWin = new Float64Array(D * F), dQ = new Float64Array(D * D), dK = new Float64Array(D * D), dV = new Float64Array(D * D);
    const dH = H.map(() => new Float64Array(D)), dQH = H.map(() => new Float64Array(D)), dKH = H.map(() => new Float64Array(D)), dVH = H.map(() => new Float64Array(D));
    for (let i = 0; i < N; i++) {
      const nb = [i, ...adj[i]], alpha = ALPHA[i], g = dZ[i];
      for (let d = 0; d < D; d++) dH[i][d] += g[d];
      const dAlpha = new Float64Array(nb.length);
      for (let k = 0; k < nb.length; k++) { const j = nb[k]; dAlpha[k] = dot(g, VH[j]); const a = alpha[k]; for (let d = 0; d < D; d++) dVH[j][d] += a * g[d]; }
      let sum = 0; for (let k = 0; k < nb.length; k++) sum += alpha[k] * dAlpha[k];
      for (let k = 0; k < nb.length; k++) {
        const j = nb[k], de = alpha[k] * (dAlpha[k] - sum) * scale;
        for (let d = 0; d < D; d++) { dQH[i][d] += de * KH[j][d]; dKH[j][d] += de * QH[i][d]; }
      }
    }
    for (let i = 0; i < N; i++) {
      outer(dQ, dQH[i], H[i], D, D); outer(dK, dKH[i], H[i], D, D); outer(dV, dVH[i], H[i], D, D);
      const back = matTvec(Q, dQH[i], D, D), back2 = matTvec(K, dKH[i], D, D), back3 = matTvec(V, dVH[i], D, D);
      for (let d = 0; d < D; d++) dH[i][d] += back[d] + back2[d] + back3[d];
      outer(dWin, dH[i], X[i], D, F);
    }
    const step = (M: Float64Array, G: Float64Array) => { const s = LR / Math.max(1, count / 40); for (let i = 0; i < M.length; i++) M[i] -= s * G[i]; };
    step(Win, dWin); step(Q, dQ); step(K, dK); step(V, dV);
  }
  forward();
  const score = (a: number, b: number) => sigmoid(dot(Z[a], Z[b]));
  // prueba ciega: AUC sobre pares retenidos frente a pares al azar
  let auc: number | null = null;
  if (holdout.length >= 3) {
    let wins = 0, total = 0;
    for (const [a, b] of holdout) for (let k = 0; k < 20; k++) { const c = Math.floor(rand() * N); if (c === a || adjSet.has(a + "|" + c)) continue; total++; if (score(a, b) > score(a, c)) wins++; }
    auc = total ? wins / total : null;
  }
  // relaciones latentes: pares no conectados con mayor puntaje (elementos, no secciones)
  const els = nodes.map((n, i) => i).filter((i) => nodes[i].kind === "element");
  const cand: LatentPair[] = [];
  const MIN = opt.minScore ?? 0.8, TOP = opt.topPairs ?? 40;
  for (let x = 0; x < els.length; x++) for (let y = x + 1; y < els.length; y++) {
    const a = els[x], b = els[y];
    if (known.has(a + "|" + b) || nodes[a].parent_id === nodes[b].parent_id) continue;
    const s = score(a, b); if (s < MIN) continue;
    const shared = adj[a].filter((j) => adj[b].includes(j) && nodes[j].kind === "element").map((j) => nodes[j].id);
    const ta = new Set(tokens(nodes[a].name).filter((t) => !STOP.has(t))), tb = tokens(nodes[b].name).filter((t) => !STOP.has(t) && ta.has(t));
    const why = [shared.length ? `${shared.length} vecino(s) en común` : "", tb.length ? `comparten «${tb.slice(0, 3).join("», «")}»` : "", `mismo ${nodes[a].facets?.data_type === nodes[b].facets?.data_type ? "tipo de dato" : "contexto de sección"}`].filter(Boolean).join("; ");
    cand.push({ a: nodes[a].id, b: nodes[b].id, score: +s.toFixed(3), shared, why: why || "cercanía en el espacio aprendido", sameSource: nodes[a].source_id === nodes[b].source_id });
  }
  cand.sort((p, q) => q.score - p.score);
  const pairs = cand.slice(0, TOP);
  // comunidades: k-medias por coseno sobre Z de elementos
  const kC = opt.communities ?? Math.max(2, Math.min(8, Math.round(Math.sqrt(els.length / 2))));
  const unit = (z: Float64Array) => { const n = Math.sqrt(dot(z, z)) || 1; const o = new Float64Array(z.length); for (let i = 0; i < z.length; i++) o[i] = z[i] / n; return o; };
  const U = new Map(els.map((i) => [i, unit(Z[i])]));
  let centers = els.slice(0, kC).map((i) => Float64Array.from(U.get(i)!));
  let assign = new Map<number, number>();
  for (let it = 0; it < 15; it++) {
    assign = new Map(els.map((i) => { let best = 0, bs = -Infinity; centers.forEach((c, k) => { const s = dot(U.get(i)!, c); if (s > bs) { bs = s; best = k; } }); return [i, best]; }));
    centers = centers.map((_, k) => { const m = els.filter((i) => assign.get(i) === k); const c = new Float64Array(D); for (const i of m) for (let d = 0; d < D; d++) c[d] += U.get(i)![d]; return m.length ? unit(c) : centers[k]; });
  }
  const communities: Community[] = centers.map((c, k) => {
    const members = els.filter((i) => assign.get(i) === k);
    const freq = new Map<string, number>();
    for (const i of members) for (const t of new Set(tokens(nodes[i].name).filter((t) => !STOP.has(t) && t.length > 2))) freq.set(t, (freq.get(t) || 0) + 1);
    const top = [...freq.entries()].sort((a, b) => b[1] - a[1]).slice(0, 3).map(([t]) => t);
    const cohesion = members.length ? members.reduce((s, i) => s + dot(U.get(i)!, c), 0) / members.length : 0;
    return { id: k, label: top.join(" · ") || `grupo ${k + 1}`, members: members.map((i) => nodes[i].id), cohesion: +cohesion.toFixed(3) };
  }).filter((c) => c.members.length >= 2).sort((a, b) => b.members.length - a.members.length);
  // flujos: fechas en orden del documento, con enlaces existentes o predichos entre pasos consecutivos
  const dates = els.filter((i) => nodes[i].facets?.data_type === "fecha");
  const secOrder = new Map(agf.nodes.filter((n) => n.kind === "section").map((n) => [n.id, n.order ?? n.number ?? 0]));
  const bySrc = new Map<string, number[]>();
  for (const i of dates) { const k = nodes[i].source_id || ""; (bySrc.get(k) || bySrc.set(k, []).get(k)!).push(i); }
  const flows: Flow[] = [];
  for (const [src, list] of bySrc) {
    list.sort((a, b) => (secOrder.get(nodes[a].section_id || "") ?? 0) - (secOrder.get(nodes[b].section_id || "") ?? 0));
    if (list.length < 3) continue;
    const steps = list.map((i, k) => { const prev = list[k - 1]; const linked = prev !== undefined && adjSet.has(prev + "|" + i); return { id: nodes[i].id, predicted: prev !== undefined && !linked, score: prev === undefined ? 1 : +score(prev, i).toFixed(3) }; });
    flows.push({ steps, label: `Secuencia de fechas en ${agf.sources.find((s) => s.id === src)?.title || agf.sources.find((s) => s.id === src)?.name || "la fuente"}` });
  }
  // reubicaciones: elemento más cercano al centro de otra sección que al de la suya
  const secCent = new Map<string, Float64Array>();
  for (const [sid] of secOrder) { const m = els.filter((i) => nodes[i].section_id === sid); if (m.length < 2) continue; const c = new Float64Array(D); for (const i of m) for (let d = 0; d < D; d++) c[d] += U.get(i)![d]; secCent.set(sid, unit(c)); }
  const relocations: Relocation[] = [];
  for (const i of els) {
    const own = nodes[i].section_id || ""; const oc = secCent.get(own); if (!oc) continue;
    const so = dot(U.get(i)!, oc); let best = own, bs = so;
    for (const [sid, c] of secCent) if (sid !== own && nodes[i].source_id === agf.nodes.find((n) => n.id === sid)?.source_id) { const s = dot(U.get(i)!, c); if (s > bs) { bs = s; best = sid; } }
    if (best !== own && bs - so > 0.08) relocations.push({ id: nodes[i].id, from: own, to: best, margin: +(bs - so).toFixed(3) });
  }
  relocations.sort((a, b) => b.margin - a.margin);
  return { pairs, communities, flows, relocations: relocations.slice(0, 20), training: { nodes: N, edges: edges.length, epochs: EPOCHS, lossStart: +lossStart.toFixed(4), lossEnd: +lossEnd.toFixed(4), holdoutAuc: auc === null ? null : +auc.toFixed(3), seed: SEED, dim: D, ms: Math.round(performance.now() - t0) } };
}

export const latentKey = (a: string, b: string) => [a, b].sort().join("|");
export const normName = norm;
